PARQUET-593: Add API for writing Page statistics

lomereiter opened the pull request.

lomereiter commented on src/parquet/column/page.h:
```diff
@@ -169,14 +170,16 @@ class PageWriter {
 
-  virtual void Close() = 0;
+  // TODO think of a better way to pass column chunk statistics
+  virtual void Close(const std::shared_ptr<RowGroupStatistics>& chunk_statistics) = 0;
```
The API is admittedly rather ugly here, because PageWriter operates on raw bytes and not values. It could handle _column chunk_ statistics by itself, but then it would have to be initialized somewhere with a concrete type.

lomereiter commented on src/parquet/column/statistics.cc:
```diff
@@ -0,0 +1,192 @@
+struct Compare {
+  explicit Compare(const ColumnDescriptor* descr)
+    : type_length_(descr->type_length()) {}
+
```
If not for `FLBAType`, I'd just define `operator<` for all the types.

lomereiter commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,133 @@
+  int64_t DistinctCount() const { return distinct_count_; }
+  int64_t NumValues() const { return num_values_; }
+
+  virtual bool HasMinMax() const = 0;
```
That's a kind of crappy substitute for C++17 `std::optional` which I avoid for the time being.

lomereiter commented on src/parquet/file/reader-internal.cc:
```diff
@@ -171,16 +171,37 @@ std::unique_ptr<PageReader> SerializedRowGroup::GetColumnPageReader(int i) {
-  result.min = &meta_data.statistics.min;
-  return result;
+std::shared_ptr<RowGroupStatistics> SerializedRowGroup::GetColumnStats(int i) const {
+  const format::ColumnMetaData& meta_data = metadata_->columns[i].meta_data;
```
This really belongs to a typed `ColumnReader`, but for now I just keep the old interface.

wesm commented on src/parquet/column/statistics.cc:
```diff
@@ -0,0 +1,185 @@
+}
+
+template <typename DType>
+void TypedRowGroupStatistics<DType>::PlainEncode(const T& src, std::string& dst) {
```
Use pointer for the out parameter (`std::string*`)

wesm commented on src/parquet/column/statistics.cc:
```diff
@@ -0,0 +1,185 @@
+}
+
+template <typename DType>
+void TypedRowGroupStatistics<DType>::PlainDecode(const std::string& src, T& dst) {
```
`T* dst`

wesm commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,127 @@
+ public:
+  int64_t NullCount() const { return null_count_; }
+  int64_t DistinctCount() const { return distinct_count_; }
+  int64_t NumValues() const { return num_values_; }
```
use `null_count()`, `distinct_count()`, etc. for these names. see https://google.github.io/styleguide/cppguide.html#Function_Names

wesm commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,127 @@
+  Type::type PhysicalType() const { return descr_->physical_type(); }
+
+ protected:
+  const ColumnDescriptor* Schema() const { return descr_; }
```
either `descr()` or `schema()` for this function name

wesm commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,127 @@
+
+  virtual ~RowGroupStatistics() {}
+
+  Type::type PhysicalType() const { return descr_->physical_type(); }
```
same naming thing here

wesm commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,127 @@
+  void Update(const T* values, int64_t num_not_null, int64_t num_null);
+
+  const T& Min() const { return min_; }
+  const T& Max() const { return max_; }
```
names

wesm commented on src/parquet/column/writer.cc:
```diff
@@ -99,13 +101,15 @@ void ColumnWriter::WriteNewPage() {
   num_buffered_values_ = 0;
   num_buffered_encoded_values_ = 0;
+  chunk_statistics_->Merge(*page_statistics_);
+  page_statistics_->Reset();
```
I've leafed through parquet-mr and parquet-format; it suggests that page_statistics_ is allowed to be empty (as you have here) but not null, if I'm reading things right (there's many layers to the onion). Is there any reason why we might want to let the page statistics be nullptr?

wesm replied:
Looking further down the review it seems that this will always be non-null, so maybe this is moot.

wesm commented on src/parquet/column/writer.h:
```diff
@@ -164,11 +177,14 @@ inline void TypedColumnWriter<DType>::WriteBatch(int64_t num_values,
   WriteValues(values_to_write, values);
 
+  auto stats = static_cast<TypedRowGroupStatistics<DType>*>(page_statistics_.get());
+  stats->Update(values, values_to_write, num_values - values_to_write);
```
Does it make sense to make writing the statistics optional? Can add an option to WriterProperties. I looked at parquet-mr and I think it always tracks the statistics; I don't have a sense of whether this will materially impact serialization performance (for applications which may not care about the column statistics).

lomereiter replied:
It makes sense indeed, I ran the benchmarks, and the performance hit is sometimes significant.

xhochy replied:
Can you post some results? I would really prefer to have them optional, at least for some columns.

lomereiter replied:
[column-io-stats.txt](https://github.com/apache/parquet-cpp/files/340188/column-io-stats.txt)
[column-io-no-stats.txt](https://github.com/apache/parquet-cpp/files/340189/column-io-no-stats.txt)

wesm commented on src/parquet/file/reader-internal.cc:
```diff
@@ -171,16 +171,37 @@ std::unique_ptr<PageReader> SerializedRowGroup::GetColumnPageReader(int i) {
+    case Type::BYTE_ARRAY:
+      return MakeColumnStats<DataType<Type::BYTE_ARRAY>>(meta_data, descr);
+    case Type::FIXED_LEN_BYTE_ARRAY:
+      return MakeColumnStats<DataType<Type::FIXED_LEN_BYTE_ARRAY>>(meta_data, descr);
```
Use the typedefs for these `DataType<T>` values? I agree it might be nicer to move this up into `TypedColumnReader` somehow and skip the switch statement, but not necessary at the moment

wesm commented on src/parquet/file/reader-internal.cc:
```diff
@@ -171,16 +171,37 @@ std::unique_ptr<PageReader> SerializedRowGroup::GetColumnPageReader(int i) {
+      return MakeColumnStats<DataType<Type::BYTE_ARRAY>>(meta_data, descr);
+    case Type::FIXED_LEN_BYTE_ARRAY:
+      return MakeColumnStats<DataType<Type::FIXED_LEN_BYTE_ARRAY>>(meta_data, descr);
+  }
```
nit: add:

```c++
default: 
  break;
```

wesm commented on src/parquet/file/writer-internal.cc:
```diff
@@ -93,7 +107,7 @@ int64_t SerializedPageWriter::WriteDataPage(int32_t num_rows, int32_t num_values
   data_page_header.__set_definition_level_encoding(ToThrift(definition_level_encoding));
   data_page_header.__set_repetition_level_encoding(ToThrift(repetition_level_encoding));
-  // TODO(PARQUET-593) statistics
+  data_page_header.__set_statistics(ToThrift(page_statistics));
```
Is it possible to avoid the extra copy of min/max (there are cases where min/max can be quite large but perhaps that is too esoteric to really be concerned about)

lomereiter replied:
Possible through manually changing `__isset.statistics` to true and operating directly on the member variable. That said, compilers might be smart enough to do copy elision.

wesm commented on src/parquet/column/statistics-test.cc:
```diff
@@ -0,0 +1,158 @@
+TYPED_TEST_CASE(TestRowGroupStatistics, TestTypes);
+
+TYPED_TEST(TestRowGroupStatistics, MinMaxEncode) {
+  this->TestMinMaxEncode();
```
Can you add some tests that attempt to verify the correctness of the comparators? Especially ByteArray/FLBA

andreweduffy replied:
Make sure when you test ByteArrayType to include some unicode characters ;)

https://github.com/apache/parquet-mr/pull/362

wesm replied:
that might be too ambitious for this patch, can we create a JIRA for it to track? C++11 now does have some built in library support for unicode (see the `<locale>` include) but it could end up being necessary to pull in ICU or some 3rd party C++ unicode library

wesm commented on src/parquet/file/file-serialize-test.cc:
```diff
@@ -96,6 +96,11 @@ class TestSerialize : public ::testing::Test {
+    auto stats = std::static_pointer_cast<Int64Statistics>(rg_reader->GetColumnStats(0));
+    ASSERT_EQ(0, stats->NullCount());
+    ASSERT_EQ(128, stats->Min());
+    ASSERT_EQ(128, stats->Max());
```
This is a fairly weak test case because there's no variety to the input data. Relates to the question around verifying the correctness of the comparators

wesm commented:
Great to make progress on this -- a few nits, but main concern is working to create more thorough test cases verifying correctness of reads and writes (since databases use these statistics for to know when to skip pages / row groups, etc.).

wesm commented:
Thank you, I will review the updated patch here in short order.

lomereiter commented:
Thanks for your time. 
I haven't got around to write comparison tests yet, but here's a quick summary of changes:
* writing statistics is disabled by default (often it makes sense only for a few specific columns)
* writing and reading is covered with tests, for single- and multi-page row groups
* all column-level settings are grouped into a struct, as it's more convenient and doesn't require multiple unordered_map accesses

wesm commented on src/parquet/column/properties.h:
```diff
@@ -85,7 +85,16 @@ static constexpr ParquetVersion::type DEFAULT_WRITER_VERSION =
+  Encoding::type encoding;
+  Compression::type codec;
+  bool collect_statistics;
+};
```
If this is strictly POD, this should be a struct (note google style guidelines on member variable names for structs vs classes)

wesm commented on src/parquet/column/properties.h:
```diff
@@ -161,26 +168,37 @@ class PARQUET_EXPORT WriterProperties {
+    Builder* statistics(bool enable = true) {
+      default_column_settings_.collect_statistics = enable;
+      return this;
+    }
```
Can you make enable not an optional argument? Seeing `...->statistics()` or `statistics('foo.bar.baz')` might not make it clear what's going on

wesm replied:
Per below, it might help to rename the method `collect_statistics()`

wesm commented on src/parquet/column/properties.h:
```diff
@@ -161,26 +168,37 @@ class PARQUET_EXPORT WriterProperties {
+    Builder* statistics(
+        const std::shared_ptr<schema::ColumnPath>& path, bool enable = true) {
+      return this->statistics(path->ToDotString(), enable);
+    }
```
ditty on these

wesm commented on src/parquet/column/properties.h:
```diff
@@ -193,46 +211,48 @@ class PARQUET_EXPORT WriterProperties {
+    return column_settings(path).codec;
+  }
+
+  bool statistics(const std::shared_ptr<schema::ColumnPath>& path) const {
```
Better name for this method?

majetideepak commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,128 @@
+
+  bool has_min_max_ = false;
+  T min_;
+  T max_;
```
The original `RowGroupStatistics` class kept the `min_`, `max_` values as pointers pointing to the actual metadata object since these values can be arbitrarily large, especially for string data. The assumption we did not document here is that the lifetime of the `RowGroupStatistics` object values is  until the `RowGroupReader`. 
But we are `copying` the `min_`, `max_` values here. @wesm any thoughts ?

wesm replied:
Thanks for catching this. 

I think there's a serious, and separate issue here because `ByteArray` and `FixedLenByteArray` do not own their data buffers. Which makes the logic in https://github.com/apache/parquet-cpp/pull/129/files#diff-59138f238492befbb8ae8a07b838463fR142 problematic

For example, between data pages in a row group, the data referenced by the pointers in this objects may be freed. This doesn't happen in the test suite (the data generated is held in a `std::vector<uint8_t>` for the duration of the test). 

So what needs to happen:

- Calling Merge with ByteArray or FLBA must copy the bytes out of the minimum/maximum object into a newly allocated `Buffer`; this should be stored in the statistics object. This way the lifetime of the value's memory can be tied locally to the lifetime of the Statistics object

- When reading a file, the RowGroupStatistics may reference the deserialized Thrift metadata in a zero-copy fashion (or, if necessary, decode the value into new memory). 

@lomereiter does this make sense? 

Practically speaking to test this, the test case will need to generate multiple batches of random data, to trigger this behavior (presumably we would get UB right now)

lomereiter replied:
Yes, thanks for pointing this subtlety out. I'll try to come up with a failing test.

wesm commented on src/parquet/column/statistics-test.cc:
```diff
@@ -0,0 +1,212 @@
+#include "parquet/file/reader.h"
+#include "parquet/file/writer.h"
+#include "parquet/util/input.h"
+#include "parquet/util/output.h"
```
Good to be in the habit of alphabetizing includes; though the whole codebase is in need of a sweep of include-what-you-use love

lomereiter replied:
In fact, latest versions of clang-format do that automatically, but I'm currently sticking to the one used on Travis CI (3.7).

wesm commented on src/parquet/column/statistics-test.cc:
```diff
@@ -0,0 +1,212 @@
+    GenerateData(num_values);
+
+    // compute statistics for the whole batch
+    ColumnDescriptor descr(node, 1 /* max definition level */, 0);
```
This is a little bit unusual; it would be better to do `const int max_def_level = 1;` if you're concerned about readability

wesm commented on src/parquet/column/statistics-test.cc:
```diff
@@ -0,0 +1,212 @@
+    ASSERT_EQ(total.max(), std::max(statistics1.max(), statistics2.max()));
+  }
+
+  using TypedStats = TypedRowGroupStatistics<TestType>;
```
Move this to the top of the fixture and use it consistently?

wesm commented on src/parquet/column/statistics-test.cc:
```diff
@@ -0,0 +1,212 @@
+    ASSERT_EQ(null_count, stats->null_count());
+    ASSERT_EQ(num_values - null_count, stats->num_values());
+    ASSERT_EQ(expected_stats.min(), stats->min());
+    ASSERT_EQ(expected_stats.max(), stats->max());
```
Cool, as discussed on the PR the last thing that's missing here is correctness of the min and max stats

wesm commented on src/parquet/file/file-serialize-test.cc:
```diff
@@ -96,6 +99,11 @@ class TestSerialize : public ::testing::Test {
+    auto stats = std::static_pointer_cast<Int64Statistics>(rg_reader->GetColumnStats(0));
+    ASSERT_EQ(0, stats->null_count());
+    ASSERT_EQ(128, stats->min());
+    ASSERT_EQ(128, stats->max());
```
Can you add a couple known extreme values to the test data?

wesm commented:
OK, outside of some nits, the two items to address are:

- Memory lifetime of the "pointer" types (ByteArray and FLBA)
- Correctness of comparators

wesm commented:
@lomereiter pinging you about this, are you able to look at this in the near future? Thanks

lomereiter commented:
@wesm yes, I'm gonna have quite long vacation soon, no worries

wesm commented:
@lomereiter can you let us know a status update on this patch when you can? I don't think it's too far away from completeness from last review

lomereiter commented:
Hi @wasm, I've been procrastinating for a while but today I tackled the main issue with the memory lifetime. I'm not sure what's the policy regarding `util/buffer.h` so I used `std::vector<uint8_t>` for storage.

wesm commented on src/parquet/column/statistics.h:
```diff
@@ -104,14 +104,15 @@ class PARQUET_EXPORT TypedRowGroupStatistics : public RowGroupStatistics {
   void PlainEncode(const T& src, std::string* dst);
   void PlainDecode(const std::string& src, T* dst);
+  void Copy(const T& src, T& dst, std::vector<uint8_t>& free_flag);
+  std::vector<uint8_t> min_buffer_, max_buffer_;
```
You can use the `OwnedMutableBuffer` here along with the associated `MemoryAllocator`

piyushnarang commented:
@lomereiter  / @wesm  - We might need to tweak stats computation for binary fields a bit based on the discussion in the PRs: https://github.com/apache/parquet-format/pull/42 and https://github.com/apache/parquet-mr/pull/362.

wesm commented:
Thanks for letting us know. What do you recommend we do right now?

piyushnarang commented:
@wesm, I think Andrew is fairly close to resolving comments on the parquet-format PR. Would be nice to ensure both the cpp & java versions are behaving in a compatible way.

wesm commented:
@lomereiter @piyushnarang looks like we should do an unsigned binary comparison here. @lomereiter it looks like comparators are the only thing left -- you should wait to rebase until PARQUET-573 is merged. thanks for patience!

wesm commented:
sorry looks like we still need to computed the signed max/min -- we can come back and add the unsigned statistics when the format version drops

https://github.com/apache/parquet-format/pull/42#discussion_r77061582

wesm commented:
@lomereiter let's proceed and compute only the signed max/min in this patch until the dust settles on parquet-format?

lomereiter commented:
@wesm ok, I'll rebase and add comparison tests this weekend so that it can be merged finally

lomereiter commented:
So far rebased only, will get back to tests and further cleanup next evening.

lomereiter commented:
Switched to using signed min/max for byte array types, added a few comparison tests. Some inefficiencies in read/write paths remain, but are topics for separate JIRAs.

wesm commented:
Awesome, will review when I can.

xhochy commented on src/parquet/column/page.h:
```diff
@@ -77,31 +80,24 @@ class DataPage : public Page {
-  friend class SerializedPageReader;
-  std::string max_;
-  std::string min_;
+  EncodedStatistics statistics_;
```
To avoid copying a lot of data, this would probably be better as a `shared_ptr`.

xhochy commented on src/parquet/column/properties.h:
```diff
@@ -82,28 +82,41 @@ static constexpr int64_t DEFAULT_PAGE_SIZE = 1024 * 1024;
 static constexpr bool DEFAULT_IS_DICTIONARY_ENABLED = true;
 static constexpr int64_t DEFAULT_DICTIONARY_PAGE_SIZE_LIMIT = DEFAULT_PAGE_SIZE;
 static constexpr int64_t DEFAULT_WRITE_BATCH_SIZE = 1024;
+static constexpr bool DEFAULT_COLLECT_STATISTICS = false;
```
Are they enabled as default in `parquet-mr`? We should probably stick to the same defaults here. Collecting statistics costs a bit but will improve query times on the files a lot.

lomereiter replied:
I checked and they are indeed enabled by default. However, if values of min/max are large (> 4KB) they are not written (https://issues.apache.org/jira/browse/PARQUET-372).

xhochy commented on src/parquet/column/properties.h:
```diff
@@ -82,28 +82,41 @@ static constexpr int64_t DEFAULT_PAGE_SIZE = 1024 * 1024;
 static constexpr Compression::type DEFAULT_COMPRESSION_TYPE = Compression::UNCOMPRESSED;
 
-using ColumnCodecs = std::unordered_map<std::string, Compression::type>;
+class PARQUET_EXPORT ColumnSettings {
```
I'm ok with this class (although it should be named `ColumnProperties`) but not with the outcome at the end. Currently it is possible to select a custom encoding for a column but fallback to the file-global defaults for all other settings. With this new class, once we have set one property on a column-basis, for this column we have to either set all other file-global defaults explicitly on this column or we'll get the parquet-cpp defaults.

lomereiter replied:
I don't understand this objection since WriterProperties::Builder interface is unchanged.

xhochy replied:
The interface is the same, just the behaviour changed.

lomereiter replied:
Gotcha. AFAIU all setters should first set `column_settings_[path]` to `default_column_settings_`.

xhochy replied:
This will still not be enough. You will also need an `isset` for each entry. I would suggest to still use the numerous maps in the builder and build the column_settings_ instances at the end.

xhochy replied:
Hope I have covered my issue in an understandable unit test: https://github.com/apache/parquet-cpp/pull/166

lomereiter replied:
Ok, went with your suggestion, thanks for putting in the test.

xhochy commented on src/parquet/column/statistics.cc:
```diff
@@ -0,0 +1,217 @@
+namespace parquet {
+
+template <typename TypedStats>
+std::shared_ptr<TypedStats> RowGroupStatistics::as() {
```
I would prefer if this was done explicitly by the user as with all other casts.

xhochy commented on src/parquet/column/statistics.cc:
```diff
@@ -0,0 +1,217 @@
+void TypedRowGroupStatistics<FLBAType>::Copy(
+    const FLBA& src, FLBA* dst, OwnedMutableBuffer& buffer) {
+  if (dst->ptr == src.ptr) return;
+  auto len = descr_->type_length();
```
Please use the explicity type here.

xhochy commented on src/parquet/column/statistics.cc:
```diff
@@ -0,0 +1,217 @@
+  auto len = descr_->type_length();
+  buffer.Resize(len);
+  for (int i = 0; i < len; i++)
+    buffer[i] = src.ptr[i];
```
Could we use `memcpy` here?

xhochy commented on src/parquet/column/statistics.cc:
```diff
@@ -0,0 +1,217 @@
+    const ByteArray& src, ByteArray* dst, OwnedMutableBuffer& buffer) {
+  if (dst->ptr == src.ptr) return;
+  buffer.Resize(src.len);
+  for (uint32_t i = 0; i < src.len; i++)
```
`memcpy`?

xhochy commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,180 @@
+
+  bool HasMinMax() const override;
+  void Reset() override;
+  void Merge(const RowGroupStatistics& other) override;
```
This function should only take `TypedRowGroupStatistics<DType>` then you could get rid of the physical type check and check at compile time.

lomereiter replied:
Thanks, moving things into TypedColumnWriter is much cleaner indeed, will do.

xhochy commented on src/parquet/column/statistics.h:
```diff
@@ -0,0 +1,180 @@
+  const T& min() const { return min_; }
+  const T& max() const { return max_; }
+
+  std::string EncodedMin() override;
```
Return const references

lomereiter replied:
Renamed to `EncodeMin`/`EncodeMax` to reflect better that these values are not stored anywhere.

xhochy commented on src/parquet/column/writer.h:
```diff
@@ -130,6 +133,18 @@ class PARQUET_EXPORT ColumnWriter {
+  std::shared_ptr<RowGroupStatistics> chunk_statistics_;
+
+  template <typename Type>
+  void InitStatistics() {
```
Please make this part of the `TypedColumnWriter`

xhochy commented on src/parquet/column/writer.cc:
```diff
@@ -172,25 +183,31 @@ void ColumnWriter::FlushBufferedDataPages() {
   }
+
+  if (properties->collect_statistics(descr_->path())) {
+    InitStatistics<Type>();
```
No need for `<Type>` if this function is part of `TypedColumnWriter`

xhochy commented on src/parquet/file/metadata.cc:
```diff
@@ -23,21 +23,49 @@
+    case Type::FIXED_LEN_BYTE_ARRAY:
+      return MakeColumnStatsT<FLBAType>(meta_data, descr);
+  }
+  return nullptr;
```
Raise an exception if the type is not supported.

xhochy commented on src/parquet/file/metadata.cc:
```diff
@@ -23,21 +23,49 @@
 namespace parquet {
 
+template <typename DType>
+static std::shared_ptr<RowGroupStatistics> MakeColumnStatsT(
```
`MakeColumnStatsTyped`

xhochy commented on src/parquet/file/reader.cc:
```diff
@@ -160,16 +160,16 @@ void ParquetFileReader::DebugPrint(
     for (auto i : selected_columns) {
       auto column_chunk = group_metadata->ColumnChunk(i);
-      const ColumnStatistics stats = column_chunk->statistics();
+      const auto stats = column_chunk->statistics();
```
Please write the type here.

xhochy replied:
@wesm Do you have a link to the guidelines at hand?

wesm replied:
https://google.github.io/styleguide/cppguide.html#auto

aside: we should survey our usages of auto to make sure we are using `const auto&` in the right places

wesm commented:
I think this is about ready to go. Needs a rebase. @xhochy ?

lomereiter commented:
Thanks all, I rebased the branch.

wesm commented:
+1
